pkg/broker/api.go

```diff
@@ -177,82 +177,83 @@ func (b *AwsBroker) Provision(request *osb.ProvisionRequest, c *broker.RequestCo
+		return nil, newHTTPStatusCodeError(http.StatusGone, "", desc)
+	}
+
+	// Get the CFN stack status
+	cfnSvc := b.Clients.NewCfn(b.GetSession(b.keyid, b.secretkey, b.region, b.accountId, b.profile, instance.Params))
+	resp, err := cfnSvc.Client.DescribeStacks(&cloudformation.DescribeStacksInput{
+		StackName: aws.String(instance.StackID),
+	})
 	if err != nil {
-		panic(err)
-	}
-	failedstates := []string{"CREATE_FAILED", "ROLLBACK_IN_PROGRESS", "ROLLBACK_FAILED", "ROLLBACK_COMPLETE", "DELETE_FAILED", "UPDATE_ROLLBACK_IN_PROGRESS", "UPDATE_ROLLBACK_FAILED", "UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS", "UPDATE_ROLLBACK_COMPLETE"}
-	progressingstates := []string{"CREATE_IN_PROGRESS", "DELETE_IN_PROGRESS", "UPDATE_IN_PROGRESS", "UPDATE_COMPLETE_CLEANUP_IN_PROGRESS"}
-	successfulstates := []string{"CREATE_COMPLETE", "DELETE_COMPLETE", "UPDATE_COMPLETE"}
-	status := *response.Stacks[0].StackStatus
-	if stringInSlice(status, failedstates) {
-		glog.Errorf("CloudFormation stack failed %#+v", si.StackID)
-		glog.Errorf(status)
-		r.LastOperationResponse.State = "failed"
-		r.LastOperationResponse.Description = response.Stacks[0].StackStatusReason
-		return &r, nil
-	} else if stringInSlice(status, progressingstates) {
-		glog.Infoln("CloudFormation stack still busy...")
-		glog.Infoln(status)
-		r.LastOperationResponse.State = "in progress"
-		r.LastOperationResponse.Description = response.Stacks[0].StackStatusReason
-		return &r, nil
-	} else if stringInSlice(status, successfulstates) {
-		glog.Infoln("CloudFormation stack operation completed...")
-		glog.Infoln(status)
-		r.LastOperationResponse.State = "succeeded"
-		r.LastOperationResponse.Description = response.Stacks[0].StackStatusReason
-		return &r, nil
+		desc := fmt.Sprintf("Failed to describe the CloudFormation stack %s: %v", instance.StackID, err)
+		return nil, newHTTPStatusCodeError(http.StatusInternalServerError, "", desc)
+	}
+
+	response := broker.LastOperationResponse{}
+	stackStatus := aws.StringValue(resp.Stacks[0].StackStatus)
+	if strings.HasSuffix(stackStatus, "_IN_PROGRESS") {
+		response.State = osb.StateInProgress
+	} else if stackStatus == cloudformation.StackStatusCreateComplete || stackStatus == cloudformation.StackStatusUpdateComplete {
+		response.State = osb.StateSucceeded
+	} else if stackStatus == cloudformation.StackStatusDeleteComplete {
+		response.State = osb.StateSucceeded
+		// If the resources were successfully deleted, try to delete the instance
+		if err := b.db.DataStorePort.DeleteServiceInstance(instance.ID); err != nil {
+			glog.Errorf("Failed to delete the service instance %s: %v", instance.ID, err)
+		}
 	} else {
-		return nil, fmt.Errorf("unexpected cfn status %v", status)
+		glog.Errorf("CloudFormation stack %s failed with status %s: %s", instance.StackID, stackStatus, aws.StringValue(resp.Stacks[0].StackStatusReason))
+		response.State = osb.StateFailed
+		response.Description = resp.Stacks[0].StackStatusReason
```

Comment:
This doesn't seem like a reliable way to get the reason for the failure. When testing, the field is often empty. But maybe that's okay for now? Alternatively, we could look through all of the stack events for failures.

Reply:
agree that it would be nice to go through the events to get a more reliable/detailed error cause, ok that this goes into a future pr